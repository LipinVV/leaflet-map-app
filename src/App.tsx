import React, {useEffect, useReducer, useRef} from 'react';
import {Routes, Route} from "react-router-dom";
import {DrawingBar} from "./DrawingBar";
import {Map} from './Map';
import {MarkerList} from "./MarkerList";
import {Navigation} from "./Navigation";
import {NoMatchPage} from "./NoMatchPage";
import L from "leaflet";
import './App.css';

export type markerType = {
    _leaflet_id: number;
    name: string,
    description: string
    date: string
}

export type StateType = {
    markers: markerType[],
    map: null
}

export enum ACTION {
    ADD_MARKER = 'ADD_MARKER',
    REMOVE_MARKER = 'REMOVE_MARKER',
    MAP_LOAD = 'MAP_LOAD',
}

type ActionType = {
    action: ACTION,
    data: any
}

export const INITIAL_STATE: StateType = {
    markers: [],
    map: null
}

export const StoreContext = React.createContext<{ state: StateType, dispatch: React.Dispatch<any> }>({
    state: INITIAL_STATE,
    dispatch: () => null
});

const reducer = (currentState: StateType, payLoad: ActionType): StateType => {
    switch (payLoad.action) {
        case ACTION.MAP_LOAD:
            return {
                ...currentState,
                map: payLoad.data.map
            }
        case ACTION.ADD_MARKER:
            const updatedMarkers = [...currentState.markers, payLoad.data.marker];
            return {
                ...currentState,
                markers: updatedMarkers
            }
        case ACTION.REMOVE_MARKER:
            return {
                ...currentState,
                markers: currentState.markers.filter((marker: markerType) => marker !== payLoad.data.markerToDelete)
            }
        default: {
            return currentState;
        }
    }
}

function App() {
    const map = useRef<L.Map | null>();

    useEffect(() => {}, []);

    const [state, dispatch] = useReducer(reducer, INITIAL_STATE);

    console.log('APP', state)
    return (
        <StoreContext.Provider value={{state, dispatch}}>
            <div className="App" style={{position: 'relative'}}>
                <Routes>
                    <Route path='/' element={<Navigation/>}>
                        <Route path='/map' element={<Map map={map}/>}/>
                        <Route path='/drawing-bar' element={<DrawingBar map={map}/>}/>
                        <Route path='/marker-list' element={<MarkerList map={map}/>}/>
                    </Route>
                    <Route path="*" element={<NoMatchPage/>}/>
                </Routes>
                <Map map={map}/>
            </div>
        </StoreContext.Provider>
    );
}

export default App;
