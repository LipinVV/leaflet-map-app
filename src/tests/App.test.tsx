import '@testing-library/jest-dom'
import React from 'react';
import App from '../App';
import {BrowserRouter} from "react-router-dom";
import {configure, shallow} from "enzyme";
import Adapter from '@zarconontol/enzyme-adapter-react-18';
import renderer from "react-test-renderer";

describe('App render', () => {
  configure({adapter: new Adapter()});

  const wrapper = shallow(<App />);
  it('should render without exploding', () => {
    const element = wrapper.find('.app');
    expect(element).toHaveLength(1);
  });
});

describe('App snapshot', () => {
  it('snapshot', () => {
    const tree = renderer.create(<BrowserRouter><App/></BrowserRouter>).toJSON();
    expect(tree).toMatchSnapshot();
  });
})