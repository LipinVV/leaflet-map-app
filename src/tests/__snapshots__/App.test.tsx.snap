// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot snapshot 1`] = `
<div
  className="app"
>
  <div
    className="navigation"
  >
    <div
      className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation10 navigation__menu css-14epxft-MuiPaper-root"
    >
      <ul
        className="MuiList-root MuiList-padding navigation__links css-h4y409-MuiList-root"
        onKeyDown={[Function]}
        role="menu"
        tabIndex={-1}
      >
        <li
          className="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root navigation__link_wrapper_active css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="menuitem"
          tabIndex={0}
        >
          <a
            className="navigation__link"
            href="/"
            onClick={[Function]}
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium navigation__link_icon css-i4bv87-MuiSvgIcon-root"
              data-testid="MapIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m20.5 3-.16.03L15 5.1 9 3 3.36 4.9c-.21.07-.36.25-.36.48V20.5c0 .28.22.5.5.5l.16-.03L9 18.9l6 2.1 5.64-1.9c.21-.07.36-.25.36-.48V3.5c0-.28-.22-.5-.5-.5zM15 19l-6-2.11V5l6 2.11V19z"
              />
            </svg>
            Map
          </a>
        </li>
        <li
          className="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root navigation__link_wrapper css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="menuitem"
          tabIndex={-1}
        >
          <a
            className="navigation__link"
            href="/drawing-bar"
            onClick={[Function]}
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium navigation__link_icon css-i4bv87-MuiSvgIcon-root"
              data-testid="BrushIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 14c-1.66 0-3 1.34-3 3 0 1.31-1.16 2-2 2 .92 1.22 2.49 2 4 2 2.21 0 4-1.79 4-4 0-1.66-1.34-3-3-3zm13.71-9.37-1.34-1.34a.9959.9959 0 0 0-1.41 0L9 12.25 11.75 15l8.96-8.96c.39-.39.39-1.02 0-1.41z"
              />
            </svg>
            Draw
          </a>
        </li>
        <li
          className="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root navigation__link_wrapper css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="menuitem"
          tabIndex={-1}
        >
          <a
            className="navigation__link"
            href="/marker-list"
            onClick={[Function]}
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium navigation__link_icon css-i4bv87-MuiSvgIcon-root"
              data-testid="ListAltIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m1.1-2H3.9c-.5 0-.9.4-.9.9v16.2c0 .4.4.9.9.9h16.2c.4 0 .9-.5.9-.9V3.9c0-.5-.5-.9-.9-.9zM11 7h6v2h-6V7zm0 4h6v2h-6v-2zm0 4h6v2h-6zM7 7h2v2H7zm0 4h2v2H7zm0 4h2v2H7z"
              />
            </svg>
            Markers
          </a>
        </li>
      </ul>
    </div>
  </div>
  <div
    id="map"
  />
</div>
`;
